packages/contracts/contracts/Router.sol

```diff
@@ -0,0 +1,220 @@
+      require(msg.value == amount, "#RC_ARF:005");
+    } else {
+      require(msg.value == 0, "#RC_ARF:006");
+      LibAsset.transferFromERC20(assetId, msg.sender, address(this), amount);
+    }
+
+    // Emit event
+    emit RelayerFeeAdded(assetId, amount, msg.sender);
+  }
+
+  function removeRelayerFee(uint256 amount, address assetId) external onlyOwner {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_RRF:002");
+
+    // Transfer funds from contract
+    LibAsset.transferAsset(assetId, payable(recipient), amount);
+
+    // Emit event
+    emit RelayerFeeRemoved(assetId, amount, msg.sender);
+  }
+
+  function removeLiquidity(
+    uint256 amount,
+    address assetId,
+    bytes calldata signature
+  ) external {
+    if (msg.sender != routerSigner) {
+      SignedRemoveLiquidityData memory payload = SignedRemoveLiquidityData({
+        amount: amount,
+        assetId: assetId,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_RL:040");
+    }
+
+    return transactionManager.removeLiquidity(amount, assetId, payable(recipient));
+  }
+
+  function prepare(
+    ITransactionManager.PrepareArgs calldata args,
+    address routerRelayerFeeAsset,
+    uint256 routerRelayerFee,
+    bytes calldata signature
+  ) external payable returns (ITransactionManager.TransactionData memory) {
+    if (msg.sender != routerSigner) {
+      SignedPrepareData memory payload = SignedPrepareData({
+        args: args,
+        routerRelayerFeeAsset: routerRelayerFeeAsset,
+        routerRelayerFee: routerRelayerFee,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_P:040");
+
+      // Send the relayer the fee
+      if (routerRelayerFee > 0) {
+        LibAsset.transferAsset(routerRelayerFeeAsset, payable(msg.sender), routerRelayerFee);
+      }
+    }
+
+    return
+      transactionManager.prepare{value: LibAsset.isNativeAsset(args.invariantData.sendingAssetId) ? msg.value : 0}(
+        args
+      );
+  }
+
+  function fulfill(
+    ITransactionManager.FulfillArgs calldata args,
+    address routerRelayerFeeAsset,
+    uint256 routerRelayerFee,
+    bytes calldata signature
+  ) external returns (ITransactionManager.TransactionData memory) {
+    if (msg.sender != routerSigner) {
+      SignedFulfillData memory payload = SignedFulfillData({
+        args: args,
+        routerRelayerFeeAsset: routerRelayerFeeAsset,
+        routerRelayerFee: routerRelayerFee,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_F:040");
+
+      // Send the relayer the fee
+      if (routerRelayerFee > 0) {
+        LibAsset.transferAsset(routerRelayerFeeAsset, payable(msg.sender), routerRelayerFee);
+      }
```

Comment:
Sanity checks

Reply:
I don't think that's necessary. This contract shouldn't have extra checks beyond the underlying contract.

Reply:
agreed ^